lib/src/component_declaration/redux_component.dart

```diff
@@ -0,0 +1,163 @@
+///   user interactions and events into [Action]s.
+/// * Redux components can use data from one or many [Store] instances to define
+///   the resulting component.
+///
+/// Use with the over_react transformer via the `@Component()` ([annotations.Component]) annotation.
+///
+/// > Related: [ReduxUiStatefulComponent]
+abstract class ReduxUiComponent<T extends ReduxUiProps> extends UiComponent<T> with _ReduxComponentMixin<T> {
+  @mustCallSuper
+  @override
+  void componentWillMount() {
+    super.componentWillMount();
+    _setUpSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillReceiveProps(Map nextProps) {
+    super.componentWillReceiveProps(nextProps);
+    // Unbind store subs so they can be re-bound in componentDidUpdate
+    // once the new props are available, ensuring the values returned by [redrawOn]
+    // are not outdated.
+    _tearDownSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentDidUpdate(Map prevProps, Map prevState) {
+    // If the subs are not bound at this point, then that means they were unbound by
+    // componentWillReceiveProps, and need to be re-bound now that new props are available.
+    if (!_areStoreSubsBound) _setUpSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillUnmount() {
+    super.componentWillUnmount();
+    _tearDownSubs();
+  }
+}
+
+/// Builds on top of [UiStatefulComponent], adding `built_redux` integration.
+///
+/// * Redux components are responsible for rendering application views and turning
+///   user interactions and events into [Action]s.
+/// * Redux components can use data from one or many [Store] instances to define
+///   the resulting component.
+///
+/// Use with the over_react transformer via the `@Component()` ([annotations.Component]) annotation.
+///
+/// > Related: [ReduxUiComponent]
+abstract class ReduxUiStatefulComponent<T extends ReduxUiProps, S extends UiState> extends UiStatefulComponent<T, S>
+    with _ReduxComponentMixin<T> {
+  @mustCallSuper
+  @override
+  void componentWillMount() {
+    super.componentWillMount();
+    _setUpSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillReceiveProps(Map nextProps) {
+    super.componentWillReceiveProps(nextProps);
+    // Unbind store subs so they can be re-bound in componentDidUpdate
+    // once the new props are available, ensuring the values returned by [redrawOn]
+    // are not outdated.
+    _tearDownSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentDidUpdate(Map prevProps, Map prevState) {
+    // If the subs are not bound at this point, then that means they were unbound by
+    // componentWillReceiveProps, and need to be re-bound now that new props are available.
+    if (!_areStoreSubsBound) _setUpSubs();
+  }
+
+  @mustCallSuper
+  @override
+  void componentWillUnmount() {
+    super.componentWillUnmount();
+    _tearDownSubs();
+  }
+}
+
+abstract class _ReduxComponentMixin<T extends ReduxUiProps> implements UiComponent<T> {
```

Comment:
I think it would make a lot of sense to have a connect function here. The default implementation could be something like

```dart
Substate connect(State state) => state;
```

But then that could be overridden to only redraw when specific pieces of your state change

```dart
@override
Substate connect(State state) => state.myComponentWantsToRedrawOn;
```

then redrawOn would be something like
```dart
List<Stream> redrawOn() {
     if (props.store is Store) {
       return <Stream>[props.store.nextSubstate(connect)];
     } else {
       return <Stream>[];
     }
  }
```

Reply:
With that how would someone listen to multiple stores?

Reply:
There is no concept of multiple stores. It is one store in redux.

Reply:
True, what about multiple values on the single store?

Reply:
Agreed. We'll want a substate on multiple values.

Reply:
So one thing you can do is map all the values into a new built_value. For example say by store's state object is:
```dart
abstract class AppState implements Built<AppState, AppStateBuilder> {
  int get foo;
  int get bar;
  int get baz;
}
```

But my component only wants foo and bar. I can still only using a single nextSubstate by defining a new model that contains the subset of state my component cares about and having my connect function return that. So i would define

```dart
abstract class DataComponentWants implements Built<DataComponentWants, DataComponentWantsBuilder> {
  int get foo;
  int get bar;
}
```

and have connect do:
```dart
DataComponentWants connect(AppState state) => new DataComponentWants(
  foo: state.foo,
  bar: state.bar,
);
```

Reply:
Since DataComponentWants is also a built, it should be comparable with the last result of connect, then we can limit redraws by only redrawing if the last result == new result